### ISSUE TITLE
DDrppi bug memory leak

### ISSUE THREAD

**Issue by Richard-Sti** [manodeep/Corrfunc#227]
## General information

- Corrfunc version: 2.3.4
- platform: Linux, macOS
- installation method (pip/source/other?): pip

## Issue description 
Each Corrfunc.theory.DDrppi call appears to be leaking memory.

### Expected behavior 
Calling DDrppi should not leave any memory inaccessible.

### Actual behavior
The amount of potentially leaked memory depends on the problem but is on order of O(100 KB) on each call. The extra allocated memory can be tracked to Corrfunc/theory/DDrppi.py:297.

In the minimal failing example if you keep running it the reserved memory will continuously increase until the program crashes.

### What have you tried so far?
Tried explicitly deleting the variables and calling the Python garbage collector.

## Minimal failing example

```python
import numpy as np
from Corrfunc.theory import DDrppi

rpbins = np.logspace(np.log10(0.1), np.log10(25), 13)
x, y, z = [np.random.uniform(0, 400, 50000) for __ in range(3)]

while True:
    __ = DDrppi(1, 1, 60, rpbins, x, y, z)
```

**Comment by lgarrison** [comment]
Thanks for the report!  I will try to reproduce it this week.

If you do `copy_particles=False`, does the leak persist?  Obviously, it should not leak in either case, though!

**Comment by Richard-Sti** [comment]
Thanks! Yes, the leak persists even if I set `copy_particles=False`.

**Comment by lgarrison** [comment]
@Richard-Sti I think I found the leak, a Python reference leak to the results structure.  I think it's present for all modules, but the rppi leak is the biggest because it's 2D.  Could you try PR #229?

**Comment by manodeep** [comment]
That is fantastic! 

@lgarrison The C-python extensions are proving to be annoying with the memory leaks - should we consider switching to something like `cffi`?

**Comment by lgarrison** [comment]
Yes!  I'm a big fan of CFFI.  I haven't used it at this scale, though!

**Comment by Richard-Sti** [comment]
@lgarrison I compiled PR #229, ran the minimal failing example and the leak looks solved now. Awesome, thanks!

**Comment by lgarrison** [comment]
Merged!  Thanks @Richard-Sti.  This will be part of the next release.

**Comment by Richard-Sti** [comment]
Great! Thanks for fixing it this quickly:)

**Comment by shanquan-gui** [comment]
@Richard-Sti
How could you fix this issue? Could you try explain it a bit further or with a python code? I have the same issue, but I do not know how to fix it .

**Comment by lgarrison** [comment]
Hi @shanquan-gui , this issue was fixed in Corrfunc v2.4.  If you're still having problems, feel free to open a new issue, and we'll be happy to help.
